web/src/saved_snippets_ui.ts

```diff
@@ -0,0 +1,110 @@
+import $ from "jquery";
+import assert from "minimalistic-assert";
+import type * as tippy from "tippy.js";
+
+import render_add_saved_snippet_modal from "../templates/add_saved_snippet_modal.hbs";
+
+import * as channel from "./channel";
+import * as compose_ui from "./compose_ui";
+import * as dialog_widget from "./dialog_widget";
+import * as dropdown_widget from "./dropdown_widget";
+import {$t_html} from "./i18n";
+import * as saved_snippets from "./saved_snippets";
+import type {StateData} from "./state_data";
+
+let saved_snippet_dropdown_widget: dropdown_widget.DropdownWidget;
+
+function submit_create_saved_snippet_form(): void {
+    const title = $<HTMLInputElement>("#add-new-saved-snippet-modal .title").val()?.trim();
+    const content = $<HTMLInputElement>("#add-new-saved-snippet-modal .content").val()?.trim();
+    if (title && content) {
+        dialog_widget.submit_api_request(channel.post, "/json/saved_snippet", {title, content});
+    }
+}
+
+function update_submit_button_state(): void {
+    const title = $<HTMLInputElement>("#add-new-saved-snippet-modal .title").val()?.trim();
+    const content = $<HTMLInputElement>("#add-new-saved-snippet-modal .content").val()?.trim();
+    const $submit_button = $("#add-new-saved-snippet-modal .dialog_submit_button");
+
+    $submit_button.prop("disabled", true);
+    if (title && content) {
+        $submit_button.prop("disabled", false);
+    }
+}
+
+function saved_snippet_modal_post_render(): void {
+    $("#add-new-saved-snippet-modal").on("input", "input,textarea", update_submit_button_state);
+}
+
+export function rerender_dropdown_widget(): void {
+    const options = saved_snippets.get_options_for_dropdown_widget();
+    saved_snippet_dropdown_widget.list_widget?.replace_list_data(options);
+}
+
+function delete_saved_snippet(saved_snippet_id: string | undefined): void {
+    assert(saved_snippet_id !== undefined);
```

Comment:
A more standard type design would be for this to require a string and have the caller do the assert.